lib/mfg.c

```diff
@@ -463,3 +485,140 @@ int switchtec_read_sec_cfg_file(FILE *setting_file,
 
 	return 0;
 }
+
+/**
+ * @brief Set KMSK entry
+ * @param[in]  dev		Switchtec device handle
+ * @param[in]  public_key	Public key
+ * @param[in]  public_key_exp	Public key exponent
+ * @param[in]  signature	Signature
+ * @param[in]  kmsk_entry_data	KMSK entry data
+ * @return 0 on success, error code on failure
+ */
+int switchtec_kmsk_set(struct switchtec_dev *dev,
+		       uint8_t *public_key,
+		       uint32_t public_key_exp,
+		       uint8_t *signature,
+		       uint8_t *kmsk_entry_data)
+{
+	int ret;
+	struct kmsk_cmd1 {
+		uint8_t subcmd;
+		uint8_t reserved[3];
+		uint8_t pub_key[SWITCHTEC_PUB_KEY_LEN];
+		uint32_t pub_key_exponent;
+	} cmd1;
+
+	struct kmsk_cmd2 {
+		uint8_t subcmd;
+		uint8_t reserved[3];
+		uint8_t signature[SWITCHTEC_SIG_LEN];
+	} cmd2;
+
+	struct kmsk_cmd3 {
+		uint8_t subcmd;
+		uint8_t num_entries;
+		uint8_t reserved[2];
+		uint8_t kmsk[SWITCHTEC_KMSK_LEN];
+	} cmd3;
+
+	if (public_key) {
+		memset(&cmd1, 0, sizeof(cmd1));
+		cmd1.subcmd = MRPC_KMSK_ENTRY_SET_PKEY;
+		memcpy(cmd1.pub_key, public_key, SWITCHTEC_PUB_KEY_LEN);
+		cmd1.pub_key_exponent = htole32(public_key_exp);
+
+		ret = switchtec_cmd(dev, MRPC_KMSK_ENTRY_SET, &cmd1,
+				    sizeof(cmd1), NULL, 0);
+		if (ret < 0)
+			return ret;
+	}
+
+	if (signature) {
+		memset(&cmd2, 0, sizeof(cmd2));
+		cmd2.subcmd = MRPC_KMSK_ENTRY_SET_SIG;
+		memcpy(cmd2.signature, signature, SWITCHTEC_SIG_LEN);
+
+		ret = switchtec_cmd(dev, MRPC_KMSK_ENTRY_SET, &cmd2,
+			    	    sizeof(cmd2), NULL, 0);
+		if (ret < 0)
+			return ret;
+	}
+
+	memset(&cmd3, 0, sizeof(cmd3));
+	cmd3.subcmd = MRPC_KMSK_ENTRY_SET_KMSK;
+	cmd3.num_entries = 1;
+	memcpy(cmd3.kmsk, kmsk_entry_data, SWITCHTEC_KMSK_LEN);
+
+	return switchtec_cmd(dev, MRPC_KMSK_ENTRY_SET, &cmd3, sizeof(cmd3),
+			     NULL, 0);
+}
+
+#if HAVE_LIBCRYPTO
+/**
+ * @brief Read public key from public key file
+ * @param[in]  pubk_file Public key file
+ * @param[out] pubk	 Public key
+ * @param[out] exp	 Public key exponent
+ * @return 0 on success, error code on failure
+ */
+int switchtec_read_pubk_file(FILE *pubk_file, uint8_t *pubk,
+			     uint32_t *exp)
+{
+	RSA	     *RSAKey = NULL;
+	const BIGNUM    *modulus_bn;
+	const BIGNUM    *exponent_bn;
+	uint32_t    exponent_tmp = 0;
+
+	RSAKey = PEM_read_RSA_PUBKEY(pubk_file, NULL, NULL, NULL);
+	if (RSAKey == NULL)
+		return -1;
+
+	RSA_get0_key(RSAKey, &modulus_bn, &exponent_bn, NULL);
+
+	BN_bn2bin(modulus_bn, pubk);
+	BN_bn2bin(exponent_bn, (uint8_t *)&exponent_tmp);
+
+	*exp = be32toh(exponent_tmp);
+	RSA_free(RSAKey);
+
+	return 0;
+}
+#endif
+
+/**
+ * @brief Read KMSK data from KMSK file
+ * @param[in]  kmsk_file KMSK file
+ * @param[out] kmsk   	 KMSK entry data
+ * @return 0 on success, error code on failure
+ */
+int switchtec_read_kmsk_file(FILE *kmsk_file, uint8_t *kmsk)
```

Comment:
Same problems as above... how big is the kmsk data? What is the user supposed to pass here?